File: 162. Find_Peak_Element/solution.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        
        int left = 0;
        int right = nums.size() - 1;
        
        if(nums.size() == 1)
            return 0;
        if(nums.size() == 2){
            if(nums[0] > nums[1])
                return 0;
            else
                return 1;
        }
            
            
        while(left < right){
            
            int mid = (left + right) / 2;
                        
            if(mid == 0){
                if(nums[mid] > nums[mid + 1])
                    return 0;
            }
            if(mid == nums.size() - 1){
                if(nums[mid] > nums[mid - 1])
                    return nums.size() - 1;
            }
            if(nums[mid - 1] < nums[mid] && nums[mid] > nums[mid + 1])
                return mid;
            
            if(nums[mid] <= nums[mid + 1])
                left = mid + 1;
            else
                right = mid;
        }
        
        return left;
    }
};

File: 162. Find_Peak_Element/solution.java
class Solution {
    public int findPeakElement(int[] nums) {
        
        int left = 0;
        int right = nums.length - 1;
        
        while(left <= right){
            
            int mid = (left + right) / 2;
            
            boolean cond1 = (mid - 1 >= 0 && nums[mid - 1] < nums[mid] && mid + 1 < nums.length && nums[mid + 1] < nums[mid]);
            boolean cond2 = (mid - 1 < 0 && mid + 1 < nums.length && nums[mid + 1] < nums[mid]);
            boolean cond3 = (mid + 1 >= nums.length && mid - 1 >= 0 && nums[mid - 1] < nums[mid]);
            boolean cond4 = (nums.length == 1);
            
            if(cond1 || cond2 || cond3 || cond4)
                return mid;
            else if(mid - 1 >= 0 && nums[mid] < nums[mid - 1])
                right = mid - 1;
            else
                left = mid + 1;
            
        }
        
        return -1;
        
    }
}